jlm/rvsdg/region.hpp

```diff
@@ -763,6 +778,40 @@ class Region
   region_bottom_node_list BottomNodes_;
   region_top_node_list TopNodes_;
   region_nodes_list Nodes_;
+  RegionObserver * observers_ = nullptr;
+
+  friend class RegionObserver;
+  friend class SimpleNode;
+  friend class StructuralNode;
+  friend class Input;
+};
+
+class RegionObserver
+{
+public:
+  ~RegionObserver();
+
+  explicit RegionObserver(Region * region);
+
+  RegionObserver(const RegionObserver &) = delete;
+
+  RegionObserver &
+  operator=(const RegionObserver &) = delete;
+
+  virtual void
+  NodeCreate(Node * node);
+
+  virtual void
+  NodeDestroy(Node * node);
+
+  virtual void
+  InputChange(Input * input, Output * old_origin, Output * new_origin);
```

Comment:
Any particular reason why they are not abstract methods?

Reply:
Defaulting them as "do nothing" adds a tiny bit of (dubious) convenience. Changing this to a clean interface.

Reply:
made it pure interface